### PR TITLE
polygon/heimdall: heimdall.Service

### PR DESCRIPTION
**Pull request by battlmonstr**
* add a new Service for the HeimdallComponent. It will replace `heimdall.go` once everything is ready.
* move entity stores from Scraper to the Service
* implement missing Heimdall interface methods

### CODE REVIEW

File: polygon/heimdall/service.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+	return libcommon.SliceMap(entities, castEntityToWaypoint), err
+}
+
+func (s *service) FetchMilestonesFromBlock(ctx context.Context, startBlock uint64) (Waypoints, error) {
+	s.scraper.Synchronize(ctx)
+	entities, err := s.milestoneStore.RangeFromBlockNum(ctx, startBlock)
+	return libcommon.SliceMap(entities, castEntityToWaypoint), err
+}
+
+func sliceTakeLast[T any](s []T, count int) []T {
```
Comment (taratorio) [inline polygon/heimdall/service.go:115]: this feels like a erigon-lib/common func to me

File: polygon/heimdall/service.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+func (s *service) FetchLatestSpan(ctx context.Context) (*Span, error) {
+	s.scraper.Synchronize(ctx)
+	entity, err := s.spanStore.GetLastEntity(ctx)
+	return downcastSpanEntity(entity), err
+}
+
+func castEntityToWaypoint(entity Entity) Waypoint {
+	return entity.(Waypoint)
```
Comment (taratorio) [inline polygon/heimdall/service.go:100]: there is a lot of unsafe casting happening...
would you address this in follow ups please?

File: polygon/heimdall/service.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+		return s[length-count:]
+	}
+	return s
+}
+
+// TODO: this limit is a temporary solution to avoid piping thousands of events
+// during the first sync. Let's discuss alternatives. Hopefully we can remove this limit.
+const maxEntityEvents = 5
+
+func (s *service) OnMilestoneEvent(callback func(*Milestone)) polygoncommon.UnregisterFunc {
```
Comment (taratorio) [inline polygon/heimdall/service.go:127]: what are your thoughts on keeping the naming consistent for the observer pattern? ie heimdallService.RegisterMilestoneObserver?

File: polygon/heimdall/service.go (hunk 1 above)
Comment (battlmonstr) [reply polygon/heimdall/service.go]: moved there

File: polygon/heimdall/service.go (hunk 2 above)
Comment (battlmonstr) [reply polygon/heimdall/service.go:100]: :+1:

File: polygon/heimdall/service.go (hunk 3 above)
Comment (battlmonstr) [reply polygon/heimdall/service.go]: renamed

**Pull request merged**